File: LFAutomation/Python/multiple_roi.py
# Import the .NET class library
import clr

# Import python sys module
import sys

# Import os module
import os

# Import System.IO for saving and opening files
from System.IO import *

# Import c compatible List and String
from System import String
from System.Collections.Generic import List

# Add needed dll references
sys.path.append(os.environ['LIGHTFIELD_ROOT'])
sys.path.append(os.environ['LIGHTFIELD_ROOT']+"\\AddInViews")
clr.AddReference('PrincetonInstruments.LightFieldViewV5')
clr.AddReference('PrincetonInstruments.LightField.AutomationV5')
clr.AddReference('PrincetonInstruments.LightFieldAddInSupportServices')

# PI imports
from PrincetonInstruments.LightField.Automation import Automation
from PrincetonInstruments.LightField.AddIns import RegionOfInterest
from PrincetonInstruments.LightField.AddIns import DeviceType

    
def set_custom_ROI():
    # Get device full dimensions
    dimensions = get_device_dimensions()
        
    regions = []
    
    # Add two ROI to regions
    regions.append(
        RegionOfInterest
        (dimensions.X, dimensions.Y,
         dimensions.Width, dimensions.Height/4,
         dimensions.XBinning, dimensions.YBinning))
    
    regions.append(
        RegionOfInterest
        (dimensions.X, dimensions.Height/2,
         dimensions.Width, dimensions.Height/4,
         dimensions.XBinning, dimensions.YBinning))

    # Set both ROI
    experiment.SetCustomRegions(regions)

    # Display the dimensions for each ROI
    for roi in regions:
        print_region(roi)    

def get_device_dimensions():
    return experiment.FullSensorRegion

def print_region(region):
    print("Custom Region Setting:")
    print(String.Format("{0} {1}", "\tX:", region.X))
    print(String.Format("{0} {1}", "\tY:", region.Y))
    print(String.Format("{0} {1}", "\tWidth:", region.Width))
    print(String.Format("{0} {1}", "\tHeight:", region.Height))
    print(String.Format("{0} {1}", "\tXBinning:", region.XBinning))
    print(String.Format("{0} {1}", "\tYBinning:", region.YBinning))

def device_found():
    # Find connected device
    for device in experiment.ExperimentDevices:
        if (device.Type == DeviceType.Camera):
            return True
     
    # If connected device is not a camera inform the user
    print("Camera not found. Please add a camera and try again.")
    return False  



# Create the LightField Application (true for visible)
# The 2nd parameter forces LF to load with no experiment 
auto = Automation(True, List[String]())

# Get experiment object
experiment = auto.LightFieldApplication.Experiment

if (device_found()==True): 
    # Set a custom region of interest
    set_custom_ROI()

    # Acquire image
    experiment.Acquire()



# Result: Two regions of interest are set in LightField.
#         They will be used to acquire an image.
#         ROI dimensions printed.
'''
Custom Region Setting:
	X: 0
	Y: 0
	Width: 1024
	Height: 64
	XBinning: 1
	YBinning: 1
Custom Region Setting:
	X: 0
	Y: 128
	Width: 1024
	Height: 64
	XBinning: 1
	YBinning: 1
'''

File: LFAutomation/MATLAB/multiple_roi.m
clear all

%n x 6 matrix, n = number of desired ROIs
%column order (Lightfield Parameters): [X Y W H B B]
regionMat=[0 0 420 504 1 1; 478 418 512 516 1 1];

Setup_LightField_Environment;
instance1=lfm(true);

exp=instance1.application.Experiment;

if device_loaded(exp)
    setROIs(instance1,regionMat);   %Custom ROIs set to whatever was in matrix

    instance1.set_frames(1);    
    data=instance1.acquire;     
    instance1.experiment.SetFullSensorRegion(); %now set to Full Frame
    dataFull=instance1.acquire;     

    %plot acquired data
    figure(1);
    subplot(1,3,1), imshow(data{1,1},[]), title('ROI 1')
    subplot(1,3,2), imshow(data{2,1},[]), title('ROI 2')
    subplot(1,3,3), imshow(dataFull,[]), title('Full Frame')

end

instance1.close;

function camLoaded=device_loaded(obj)
    
    camLoaded=0;
    devList=obj.ExperimentDevices;
    objects=devList.Count;

    for i=1:objects
        device=devList.Item(i-1);
        type=string(device.Type);
        if strcmp(type,'Camera')
            camLoaded=1;
            break;
        end
    end
    if ~camLoaded
        fprintf('Camera not found. Please add a camera and try again.\n');
    end
end


function setROIs(obj,regionMat)      
regions=size(regionMat,1);
regionArray = NET.createArray('PrincetonInstruments.LightField.AddIns.RegionOfInterest',regions);
for i=1:regions
    x=regionMat(i,1); y=regionMat(i,2); width=regionMat(i,3);
    height=regionMat(i,4); xb=regionMat(i,5); yb=regionMat(i,6);
    regionArray(i) = PrincetonInstruments.LightField.AddIns.RegionOfInterest(x,y,width,height,xb,yb);
end

obj.experiment.SetCustomRegions(regionArray);

end
